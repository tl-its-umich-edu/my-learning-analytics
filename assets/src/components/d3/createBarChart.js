import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { margin } from '../../constants/chartConstants'

function createBarChart ({ data, width, height, el, tip }) {
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  const x = d3.scaleBand()
    .domain(data.map(d => d.label))
    .range([margin.left, aWidth - margin.right])
    .padding(0.1)

  const y = d3.scaleLinear()
    .domain([0, d3.max(data, d => d.data)]).nice()
    .range([aHeight - margin.bottom, margin.top])

  const svg = d3.select(el).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const bar = svg.selectAll('.bar')
    .data(data).enter()
    .append('rect')
    .attr('class', 'bar')
    .attr('x', d => x(d.label))
    .attr('width', x.bandwidth())
    .attr('y', d => y(d.data))
    .attr('height', d => y(0) - y(d.data))

  svg.append('g')
    .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
    .call(d3.axisBottom(x))

  svg.append('g')
    .attr('transform', `translate(${margin.left},0)`)
    .call(d3.axisLeft(y))

  if (tip) {
    svg.call(tip)
    bar
      .on('mouseover', tip.show)
      .on('mouseout', tip.hide)
  }
}

export default createBarChart
