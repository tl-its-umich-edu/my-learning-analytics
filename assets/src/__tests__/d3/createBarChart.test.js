/* global describe, test, expect */
import createBarChart from '../../components/d3/createBarChart'
import { barChartData } from '../testData/d3TestData'

describe('createBarChart', () => {
  test('should build a bar chart', () => {
    const div = document.createElement('div')
    createBarChart({ data: barChartData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
