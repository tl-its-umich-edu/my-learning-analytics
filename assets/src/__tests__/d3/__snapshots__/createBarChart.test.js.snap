// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBarChart should build a bar chart 1`] = `
<div>
  <svg
    height="450"
    width="940"
  >
    <rect
      class="bar"
      height="60"
      width="111.54929577464789"
      x="52.3943661971831"
      y="360"
    />
    <rect
      class="bar"
      height="170"
      width="111.54929577464789"
      x="176.33802816901408"
      y="250"
    />
    <rect
      class="bar"
      height="390"
      width="111.54929577464789"
      x="300.28169014084506"
      y="30"
    />
    <rect
      class="bar"
      height="115"
      width="111.54929577464789"
      x="424.22535211267603"
      y="305"
    />
    <rect
      class="bar"
      height="50"
      width="111.54929577464789"
      x="548.169014084507"
      y="370"
    />
    <rect
      class="bar"
      height="220.00000000000003"
      width="111.54929577464789"
      x="672.112676056338"
      y="199.99999999999997"
    />
    <rect
      class="bar"
      height="20"
      width="111.54929577464789"
      x="796.056338028169"
      y="400"
    />
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 420)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H920.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(108.16901408450704,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Bob
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(232.11267605633802,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Robin
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(356.056338028169,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Anne
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(480,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Mark
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(603.9436619718309,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Joe
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(727.887323943662,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Eve
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(851.8309859154929,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Karen
        </text>
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,420.5H0.5V20.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,420.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,370.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,320.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,270.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          30
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,220.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,170.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          50
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,120.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,70.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          70
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,20.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          80
        </text>
      </g>
    </g>
  </svg>
</div>
`;
